### PR TITLE
[as5916-54xks] Modify IDPROM size/Add 1GBase-T SFP sysfs/Update thermal threshold

### PR DESCRIPTION
**Pull request by brandonchuang**

### CODE REVIEW

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c [hunk 1]
```diff
@@ -1342,6 +1369,111 @@ static ssize_t set_qsfp_lpmode(struct device *dev, struct device_attribute *da,
...
+{
+    int i = 0;
+	int status;
+    char *token = NULL, *last = NULL, *delim = " ";
+    struct sfp_phy_write_data wdata;
+    char command[NUM_OF_PHY_REGISTERS * 3 + 3] = {0};
+    char *ipmi_phy_tx_data = (char *)&wdata;
+
+    memset(&wdata, 0, sizeof(wdata));
+    memcpy(command, buf, sizeof(command));
```
Comment (billie-alsup) [inline packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: Does this run the risk of a fault if "buf" is near the end of a page, there is no adjacent page, and (buf + sizeof(command)) runs past the end of the page?  Even if there is no fault, should you be copying arbitrary kernel data into your command buffer here?  Would it not be better to take the minimum of sizeof(command) and count?  If count exceeds sizeof(command), should that be reflected in an error return?

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c [hunk 2]
```diff
@@ -1342,6 +1369,111 @@ static ssize_t set_qsfp_lpmode(struct device *dev, struct device_attribute *da,
...
+    token = __strtok_r(command, delim, &last);
+    while (token != NULL) {
+        long param = 0;
+
+    	status = kstrtol(token, 0, &param);
+    	if (status) {
+    		return status;
+    	} 
+
+        ipmi_phy_tx_data[i++] = param;
```
Comment (billie-alsup) [inline packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: should you not have a range check here, and return an error for an out of range parameter?

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c [hunk 3]
```diff
@@ -1342,6 +1369,111 @@ static ssize_t set_qsfp_lpmode(struct device *dev, struct device_attribute *da,
...
+    if (unlikely(status != 0)) {
+        goto exit;
+    }
+
+    if (unlikely(data->ipmi.rx_result != 0)) {
+        status = -EIO;
+        goto exit;
+    }
+
+    status = count;
```
Comment (billie-alsup) [inline packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: this seems incorrect, considering you will at most process sizeof(command) bytes.

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c (hunk 1 above)
Comment (brandonchuang) [reply packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: There is a logic error here.
I should declare a string buffer "char *command "then copy from the "const char *buf" passed into set_phy()

I modified as below
```suggestion
    char *command = NULL;
    char *ipmi_phy_tx_data = (char *)&wdata;

    command = kzalloc(count+1, GFP_KERNEL);
    if (!command) {
        return -ENOMEM;
    }

    memset(&wdata, 0, sizeof(wdata));
    memcpy(command, buf, count);
```

The variable "const char *buf" is guaranteed to be safe to access with the length "count" by sysfs core, the buf is in kernel space for driver to access, the copy_from_user is also hidden in core.
Here is the code slice in file.c (/fs/sysfs/file.c)
```suggestion
static int 
fill_write_buffer(struct sysfs_buffer * buffer, const char __user * buf, size_t count)
{
	int error;

	if (!buffer->page)
		buffer->page = (char *)get_zeroed_page(GFP_KERNEL);
	if (!buffer->page)
		return -ENOMEM;

	if (count >= PAGE_SIZE)
		count = PAGE_SIZE - 1;
	error = copy_from_user(buffer->page,buf,count);
```

So the memcpy from buf is safe.

Please review my changes again.

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c (hunk 2 above)
Comment (brandonchuang) [reply packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: Done for adding the range check.

```suggestion
        /* Validate each param length */
        if (param > 0xFF) {
            return -EINVAL;
        }
```

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c (hunk 3 above)
Comment (brandonchuang) [reply packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: The range is already verified here, the original code is expected.

```suggestion
    /* Validate command */
    if (i <= IPMI_PHY_HEADER_LEN || (i != IPMI_PHY_DATA_LEN(wdata.ipmi_tx_data[2]))) {
        return -EINVAL;
    }
```

File: packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c (hunk 2 above)
Comment (brandonchuang) [reply packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c]: Fix the command leak and range checking.
Please help verify.

**Pull request merged**